### DISCUSSION TITLE
React Native: `Property 'Document' doesn't exist`

### DISCUSSION THREAD

**Discussion by rodperottoni** [mswjs/msw#2359] [Q&A] [answered]
I'm currently using `msw@2.4.11` on Expo SDK 51. When calling `.listen()` on my `server` instance, all requests start failing with the following error:

```
[ReferenceError: Property 'Document' doesn't exist]
```

- My `setupServer` is indeed imported from `msw/native`.
- My polyfills are correctly set up

<img width="469" alt="Screenshot 2024-11-15 at 3 48 31 pm" src="https://github.com/user-attachments/assets/607f7ee7-9d7d-4493-ad8a-02f5294b09ed">

After digging around, I found out the latest `XMLHttpRequestInterceptor` makes use of the global `Document` property, which is not available on React Native or Expo applications. 

The usage can be seen here: [XMLHttpRequestController.ts](https://github.com/mswjs/interceptors/blob/main/src/interceptors/XMLHttpRequest/XMLHttpRequestController.ts)

I'm wondering if anyone else has faced this issue and how did they solve it?

**Comment by rodperottoni** [comment]
Some more information:
- I've placed several console.logs around the compiled MSW code, and found out that the imported `XMLHttpRequestInterceptor` somehow is resolved to the `browser` variant under `@mswjs/interceptors`.

<img width="744" alt="Screenshot 2024-11-19 at 10 53 15 am" src="https://github.com/user-attachments/assets/238dee2e-8f50-4fd9-b70b-58a2349cb0e3">

---- 

- I can confirm this by placing logs in the following file: `node_modules/@mswjs/interceptors/lib/browser/interceptors/XMLHttpRequest/index.js`

<img width="755" alt="Screenshot 2024-11-19 at 10 54 10 am" src="https://github.com/user-attachments/assets/198b1b40-3425-4047-8ae8-3b6661f62663">

---- 

**Output:**
<img width="93" alt="Screenshot 2024-11-19 at 10 55 20 am" src="https://github.com/user-attachments/assets/43e06fb3-e869-42e8-bb43-c8fe4e731078">

- This is how I'm setting up my server:

<img width="830" alt="Screenshot 2024-11-19 at 10 57 19 am" src="https://github.com/user-attachments/assets/dbe7de20-ce90-4c5c-834b-939eca32c05a">

----

Relevant packages:
- `msw@2.4.11`
- `detox@20.25.5`
- `expo@51.0.32`
- `react-native@0.75.3`

**Reply by rodperottoni** [reply]
I think I just found my answer.

In Expo's metro config, the `config.resolver.resolverMainFields` has a default value of: https://github.com/expo/expo/blob/58c7c9156e86dab9e3f2f15135e52f03880b6cf3/packages/%40expo/metro-config/src/ExpoMetroConfig.ts#L262

For reference, this is Metro's docs on `resolverMainFields`: https://metrobundler.dev/docs/configuration/#resolvermainfields

This makes me wonder: how did anyone else ever got MSW to work with React Native + Expo if the module resolution is using the `browser` exports rather than `node` ?

**Reply by watadarkstar** [reply]
@rodperottoni Did you ever figure out how to get MSW to work with an Expo app? If so, care to share some code?

**Reply by rodperottoni** [reply]
@watadarkstar unfortunately not. I ended up putting this endeavour on the back burner.

**Reply by watadarkstar** [reply]
@rodperottoni Thanks for the quick response! I think I will also back burner it

**Comment by kettanaito** [comment, marked as answer]
I will never understand why project just opt-in into using seemingly incompatible export conditions by default. Since Metro prioritizes `browser`, you get the browser build of `@mswjs/interceptors`, which includes XHR interception as it should.

> This makes me wonder: how did anyone else ever got MSW to work with React Native + Expo

We have been looking for a champion to bring the React Native integration into a good shape. None has showed up as of now. Since not many issues have been reported about this, I deem React Native support as a non-priority task. I don't develop in React Native so it's hard for me to have a proper grasp on the workflows and project specifics there. 

## Solution

The solution is to configure `config.resolver.resolverMainFields` to prioritize `node` **only when testing**. Your Metro app may be okay digesting browser packages, but your test runs in Node.js, and it must _never_ import a browser package there.

**Reply by tranhl** [reply]
@kettanaito FWIW React Native has experimental support for the `react-native` export condition. Perhaps that's something that `msw` could leverage for it's integration? Perhaps the integration could also be marked experimental as well seeing how unstable it currently is? That would open up a path to stability that lines up with package exports support in React native as well.

More details: https://metrobundler.dev/docs/package-exports/#replacing-browser-and-react-native-fields

**Comment by j3lev** [comment]
I was able to resolve this issue by switching my HTTP client to Fetch vs. Axios. Luckily we were early enough in our project where this wasn't a huge lift. It might make sense to officially document the lack of Axios support. Interestingly enough I was still seeing console errors after switching to Fetch but the responses were resolving correctly. I didn't have time to debug further and just ended up suppressing the error.
